File: app/src/main/java/org/eyeseetea/malariacare/presentation/viewmodels/Observations/ActionViewModel.java
package org.eyeseetea.malariacare.presentation.viewmodels.Observations;

import java.util.Date;

public class ActionViewModel {
    private  String activityAction;
    private  String subActivityAction;
    private  Date dueDateAction;
    private  String responsibleAction;
    private  boolean realized;

    public ActionViewModel(
            String activityAction ,String subActivityAction, Date dueDateAction,
            String responsibleAction, boolean realized) {
        this.activityAction = activityAction;
        this.subActivityAction = subActivityAction;
        this.dueDateAction = dueDateAction;
        this.responsibleAction = responsibleAction;
        this.realized = realized;
    }

    public String getActivityAction() {
        return activityAction;
    }

    public void setActivityAction(String activityAction) {
        this.activityAction = activityAction;
    }

    public String getSubActivityAction() {
        return subActivityAction;
    }

    public void setSubActivityAction(String subActivityAction) {
        this.subActivityAction = subActivityAction;
    }

    public Date getDueDateAction() {
        return dueDateAction;
    }

    public void setDueDateAction(Date dueDateAction) {
        this.dueDateAction = dueDateAction;
    }

    public String getResponsibleAction() {
        return responsibleAction;
    }

    public void setResponsibleAction(String responsibleAction) {
        this.responsibleAction = responsibleAction;
    }

    public boolean isRealized() {
        return realized;
    }

    public void setRealized(boolean realized) {
        this.realized = realized;
    }
}

File: app/src/main/java/org/eyeseetea/malariacare/presentation/viewmodels/observations/ActionViewModel.kt
package org.eyeseetea.malariacare.presentation.viewmodels.observations

import java.util.Date

data class ActionViewModel(
    var description: String = "",
    var dueDate: Date?,
    var responsible: String = "",
    private val completionDateField: Date?
) {

    var completionDate: Date? = null
        private set

    var isCompleted: Boolean
        get() = completionDate != null
        set(completed) = if (completed) {
            completionDate = Date()
        } else {
            completionDate = null
        }

    val isEmpty: Boolean
        get() = description.isEmpty() && dueDate == null && responsible.isEmpty()

    val isValid: Boolean
        get() = description.isNotEmpty() && dueDate != null && responsible.isNotEmpty()

    init {
        this.completionDate = completionDateField
    }
}
